### ISSUE TITLE
Travis failing for pull requests

### ISSUE THREAD

**Issue by JohnStott** [dmlc/xgboost#1819]
Failing with this error message:

The command "tests/travis/run_test.sh" exited with 255.

**Comment by AbdealiJK** [comment]
This error is too generic, there must be many more meaningful errors on top of it.
The error message you've mentioned just says "The tests failed" which doesn't help...

Either a link to the failed travis or a complete error message would help debug it.

**Comment by JohnStott** [comment]
Fair enough comment.  I was just testing the waters to see if this was a known issue.  Seems to be affecting all pull requests so thought a head project programmer may already know about this.  I'll investigate in a little while to find exact error...

**Comment by JohnStott** [comment]
Okay it seems to be failing here:

test_with_pandas.py
This is the error I get (I've highlighted where the error appears to stem from):
```

[08:27:55] dmlc-core/include/dmlc/logging.h:235: [08:27:55] src/io/local_filesys.cc:86: LocalFileSystem.ListDirectory demo/data error: No such file or directory
EException AttributeError: "'DMatrix' object has no attribute 'handle'" in <bound method DMatrix.__del__ of <xgboost.core.DMatrix object at 0x000000000A8CAB70>> ignored
   B  A_X  A_Y  A_Z
0  1    1    0    0
1  2    0    1    0
2  3    0    0    1
['int', 'float', 'float', 'float']
F.
======================================================================
ERROR: test_cv_as_pandas (__main__.TestPandas)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_with_pandas.py", line 102, in test_cv_as_pandas
    dm = xgb.DMatrix(dpath + 'agaricus.txt.train')
  File "C:\Users\xyz\Anaconda2\lib\site-packages\xgboost-0.6-py2.7.egg\xgboost\core.py", line 262, in __init__
    ctypes.byref(self.handle)))
  File "C:\Users\xyz\Anaconda2\lib\site-packages\xgboost-0.6-py2.7.egg\xgboost\core.py", line 127, in _check_call
    raise XGBoostError(_LIB.XGBGetLastError())
XGBoostError: [08:27:55] src/io/local_filesys.cc:86: LocalFileSystem.ListDirectory demo/data error: No such file or directory

======================================================================

FAIL: test_pandas (__main__.TestPandas)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_with_pandas.py", line 73, in test_pandas
    assert dm.feature_types == ['int', 'int', 'int', 'int']
AssertionError

----------------------------------------------------------------------
Ran 3 tests in 0.050s

FAILED (failures=1, errors=1)
```

**Comment by JohnStott** [comment]
Seems this python code is failing:

 ```
   df = pd.DataFrame({'A': ['X', 'Y', 'Z'], 'B': [1, 2, 3]})
        dummies = pd.get_dummies(df)
        #    B  A_X  A_Y  A_Z
        # 0  1    1    0    0
        # 1  2    0    1    0
        # 2  3    0    0    1
        result, _, _ = xgb.core._maybe_pandas_data(dummies, None, None)
        exp = np.array([[1., 1., 0., 0.],
                        [2., 0., 1., 0.],
                        [3., 0., 0., 1.]])
        np.testing.assert_array_equal(result, exp)
        dm = xgb.DMatrix(dummies)
        assert dm.feature_names == ['B', 'A_X', 'A_Y', 'A_Z']
        assert dm.feature_types == ['int', 'int', 'int', 'int']

```

dm.feature_types is actually returning ['int', 'float', 'float', 'float'] which is incorrect.  Please can anyone (AbdealiJK) try this on a fresh version of xgboost?  See if you get same error?

**Comment by JohnStott** [comment]
It appears from first glance that this is a purely python issue: this function attempts to get feature types in the python DMatrix constructor:

https://github.com/dmlc/xgboost/blob/master/python-package/xgboost/core.py#L168

I'll investigate further...

**Comment by JohnStott** [comment]
Okay think I have found the problem, I have simplified here what is happening to this code:
```
        df = pd.DataFrame({'A': ['X', 'Y', 'Z'], 'B': [1, 2, 3]})
        dummies = pd.get_dummies(df)
        data_dtypes = dummies.dtypes
        print dummies
        print data_dtypes
```

```
B  A_X  A_Y  A_Z
0  1    1    0    0
1  2    0    1    0
2  3    0    0    1

B        int64
A_X    float64
A_Y    float64
A_Z    float64
dtype: object
```

Perhaps somebody can comment on whether this assertion check needs removing or what we should do?

**Comment by JohnStott** [comment]
I am using pandas version 17 and found this pandas change:

http://pandas.pydata.org/pandas-docs/version/0.19.1/whatsnew.html#get-dummies-now-returns-integer-dtypes

Maybe travis needs to be using latest pandas version if not already?  I'll install version 19 and see if this corrects problem?

**Comment by JohnStott** [comment]
Assertion works fine with 0.19.1 so I think that is the solution.  Travis needs to be using latest version of pandas!

**Comment by AbdealiJK** [comment]
On my local pandas 0.19 it works fine.

Also, I do have travis PRs which don't fail on the pandas tests - https://github.com/dmlc/xgboost/pull/1829

Are you sure you have rebased on master ? Maybe the fix has been merged in master already but not in your PR ?

**Comment by JohnStott** [comment]
AbdealiJK - please re-read my posts in particular:
>**Assertion works fine with 0.19.1** so I think that is the solution. Travis needs to be using **latest** version of pandas!

**Comment by JohnStott** [comment]
So in summary, just to clarify if not clear, when tested locally using Pandas 0.17 it failed, when I updated to 0.19.1 it worked fine.  So the assumption is that Travis is not using latest pandas version (it may now, nobody has mentioned whether this has been updated?)  Or maybe we can somehow find what version Travis is using?  Regardless, with your unit testing, I will cancel my pull requests and use your unit tests. Just need to find the time to edit and re-version git etc.

**Comment by JohnStott** [comment]
just to re-emphasise - someone else's pull request is failing with same error:

https://travis-ci.org/dmlc/xgboost/jobs/180733537

**Comment by AbdealiJK** [comment]
@JohnStott I did read all your posts, and suggested a way to test whether it was fixed - i.e. just to update your PR by rebasing it on master ...

Regarding the travis log you mentioned in https://github.com/dmlc/xgboost/issues/1819#issuecomment-264664346:
Here is the Travis pandas version is 0.191 - [check L626](https://travis-ci.org/dmlc/xgboost/jobs/180733537#L626)
Note that the link youAlso, it does not mention a sklear error, it has an error in ` xgboost/tests/python/test_basic_models.py`. So, this travis log does not seem to be related to this discussion.

In your currently open PR at https://github.com/dmlc/xgboost/pull/1710 (with travis at [179403614](https://travis-ci.org/dmlc/xgboost/jobs/179403614)) the pandas error does not seem to be appearing again. (Note that this is also using [Pandas 0.19.1](https://travis-ci.org/dmlc/xgboost/jobs/179403614#L600))

So, all in all:
 - I don't seem to be able to see this on any of the recent travis builds.
 - The recent travis builds are using pandas 0.19.1

So, it was either a secondary problem or the problem has already been fixed.
I hope this helped.

**Comment by AbdealiJK** [comment]
Just realized that you recently pushed commits on your PR at https://github.com/dmlc/xgboost/pull/1710 which is probably why the travis error is not the same anymore.

Cool, nice to see it worked :+1:
I think this issue can be closed ?

**Comment by JohnStott** [comment]
I will rebase again... see if it is now fixed...be back in half an hour...

**Comment by JohnStott** [comment]
Hmmm, seems to have failed again...  Thank you for your help so far...
It is unfortunate that the 'The command "tests/travis/run_test.sh" exited with 255.' error doesn't contain any more detail...  Please have a look at my latest merge attempt link:
https://travis-ci.org/dmlc/xgboost/jobs/181080162L777
it doesn't mention the pandas package at all (unlike the link you posted)?

Maybe this is failing somewhere else?  Perhaps the pandas 0.19 issue has been resolved but now something else is failing?

I have an hour free so I'm going to cancel my pull request, and submit a simple xgboost fix that is in desperate need of being merged, see if this causes this dreaded 255 error?

**Comment by AbdealiJK** [comment]
@JohnStott In the link you've given ([link](https://travis-ci.org/dmlc/xgboost/jobs/181080162L777)), the error can be seen at:
```
tests/python/test_api_internals.py:6:1: E302 expected 2 blank lines, found 1
```
flake8 says it was expecting 2 lines instead of 1 in the file mentioned there in accordance with PEP guidelines for python

Also, the reason pandas is not there in this travis is because it is the `TASK=python_lightweight_task` which is supposed to run without pandas, sklearn, etc.

**Comment by JohnStott** [comment]
Thanks AbdealiJK.  It seems this was a number of issues, Travis was originally giving me this error:

python-package/xgboost/rabit.py:135:1: E305 expected 2 blank lines after class or function definition, found 1
python-package/xgboost/rabit.py:208:1: E305 expected 2 blank lines after class or function definition, found 1
as per (https://travis-ci.org/dmlc/xgboost/jobs/176671938)

but this was nothing to do with my submit, I was therefore thinking this stemmed from somewhere else, something I had done (as I wouldn't expected issues to get merged)!  I.e., I then tested the python "checks" locally, for which I found the above error with regards to Pandas, which I can now see was a "red herring".  It seems in the interim, this "rabit" issue was fixed, but then my python code was then failing as per above (which I missed):
tests/python/test_api_internals.py:6:1: E302 expected 2 blank lines, found 1

I have just submitted a different pull request and it passed all checks fine.  Many thanks for your help and apologies for any inconvenience.  I will now close this.
